packages/core-contracts/contracts/token/ERC721.sol

```diff
@@ -0,0 +1,232 @@
+            revert Unauthorized(msg.sender);
+        }
+
+        store.tokenApprovals[tokenId] = to;
+
+        emit Approval(holder, to, tokenId);
+    }
+
+    function getApproved(uint256 tokenId) external view virtual override returns (address) {
+        if (!_exists(tokenId)) {
+            revert InvalidTokenId(tokenId);
+        }
+
+        return _erc721Store().tokenApprovals[tokenId];
+    }
+
+    function setApprovalForAll(address operator, bool approved) external virtual override {
+        if (msg.sender == operator) {
+            revert InvalidOperator(operator);
+        }
+
+        _erc721Store().operatorApprovals[msg.sender][operator] = approved;
+        emit ApprovalForAll(msg.sender, operator, approved);
+    }
+
+    function isApprovedForAll(address holder, address operator) public view virtual override returns (bool) {
+        return _erc721Store().operatorApprovals[holder][operator];
+    }
+
+    function transferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public payable virtual override {
+        _transfer(from, to, tokenId);
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public payable virtual override {
+        safeTransferFrom(from, to, tokenId, "");
+    }
+
+    function safeTransferFrom(
+        address from,
+        address to,
+        uint256 tokenId,
+        bytes memory data
+    ) public payable virtual override {
+        _transfer(from, to, tokenId);
+        if (!_checkOnERC721Received(from, to, tokenId, data)) {
+            revert InvalidTransferRecipient(to);
+        }
+    }
+
+    function _mint(address to, uint256 tokenId) internal virtual {
+        ERC721Store storage store = _erc721Store();
+        if (to == address(0)) {
+            revert InvalidTo(to);
+        }
+
+        if (_exists(tokenId)) {
+            // Already minted
+            revert InvalidTokenId(tokenId);
+        }
+
+        store.balanceOf[to] += 1;
+        store.ownerOf[tokenId] = to;
+
+        emit Transfer(address(0), to, tokenId);
+    }
+
+    function _burn(uint256 tokenId) internal virtual {
+        ERC721Store storage store = _erc721Store();
+        address holder = store.ownerOf[tokenId];
+
+        store.tokenApprovals[tokenId] = address(0);
+        emit Approval(holder, address(0), tokenId);
+
+        store.balanceOf[holder] -= 1;
+        delete store.ownerOf[tokenId];
+
+        emit Transfer(holder, address(0), tokenId);
+    }
+
+    function _exists(uint256 tokenId) internal view virtual returns (bool) {
+        return _erc721Store().ownerOf[tokenId] != address(0);
+    }
+
+    function _transfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual {
+        ERC721Store storage store = _erc721Store();
+
+        // Not checking tokenId existence since is checked in ownerOf()
+
+        if (!_isAuthorizedToTransfer(from, tokenId)) {
```

Comment:
easy to read than the previous version and extracting the check will ease the implementation of next PR (system owned)

Reply:
Moved to the interface calls